libs/uuid/uuid.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` 
.    ds C' 
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "UUID 1"
.TH UUID 1 "OSSP uuid 1.6.2" "04-Jul-2008" "Universally Unique Identifier"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
\&\fB\s-1OSSP\s0 uuid\fR \- \fBUniversally Unique Identifier Command-Line Tool\fR
.SH "VERSION"
.IX Header "VERSION"
\&\s-1OSSP\s0 uuid \s-11.6.2 (04-Jul-2008)\s0
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBuuid\fR
[\fB\-v\fR \fIversion\fR]
[\fB\-m\fR]
[\fB\-n\fR \fIcount\fR]
[\fB\-1\fR]
[\fB\-F\fR \fIformat\fR]
[\fB\-o\fR \fIfilename\fR]
[\fInamespace\fR \fIname\fR]
.PP
\&\fBuuid\fR
\&\fB\-d\fR
[\fB\-r\fR]
[\fB\-o\fR \fIfilename\fR]
\&\fIuuid\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fB\s-1OSSP\s0 uuid\fR is a \s-1ISO\-C:1999\s0 application programming interface (\s-1API\s0) and
corresponding command line interface (\s-1CLI\s0) for the generation of \s-1DCE\s0
1.1, \s-1ISO/IEC\s0 11578:1996 and \s-1IETF\s0 \s-1RFC\-4122\s0 compliant \fIUniversally Unique
Identifier\fR (\s-1UUID\s0). It supports \s-1DCE\s0 1.1 variant UUIDs of version 1 (time
and node based), version 3 (name based, \s-1MD5\s0), version 4 (random number
based) and version 5 (name based, \s-1SHA\-1\s0). Additional \s-1API\s0 bindings are
provided for the languages \s-1ISO\-\*(C+:1998\s0, Perl:5 and \s-1PHP:4/5\s0. Optional
backward compatibility exists for the ISO-C \s-1DCE\-1\s0.1 and Perl Data::UUID
APIs.
.PP
UUIDs are 128 bit numbers which are intended to have a high likelihood
of uniqueness over space and time and are computationally difficult
to guess. They are globally unique identifiers which can be locally
generated without contacting a global registration authority. UUIDs
are intended as unique identifiers for both mass tagging objects
with an extremely short lifetime and to reliably identifying very
persistent objects across a network.
.PP
This is the command line interface (\s-1CLI\s0) of \fB\s-1OSSP\s0 uuid\fR. For a
detailed description of UUIDs see the documentation of the application
programming interface (\s-1API\s0) in \fIuuid\fR\|(3).
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-v\fR \fIversion\fR" 3
.IX Item "-v version"
Sets the version of the generated \s-1DCE\s0 1.1 variant \s-1UUID\s0. Supported
are \fIversion\fR "\f(CW1\fR\*(L", \*(R"\f(CW3\fR\*(L", \*(R"\f(CW4\fR\*(L" and \*(R"\f(CW5\fR\*(L". The default is \*(R"\f(CW1\fR".
.Sp
For version 3 and version 5 UUIDs the additional command line arguments
\&\fInamespace\fR and \fIname\fR have to be given. The \fInamespace\fR is either
a \s-1UUID\s0 in string representation or an identifier for internally
pre-defined namespace UUIDs (currently known are "\f(CW\*(C`ns:DNS\*(C'\fR\*(L",
\&\*(R"\f(CW\*(C`ns:URL\*(C'\fR\*(L", \*(R"\f(CW\*(C`ns:OID\*(C'\fR\*(L", and \*(R"\f(CW\*(C`ns:X500\*(C'\fR"). The \fIname\fR is a string of
arbitrary length.
.IP "\fB\-m\fR" 3
.IX Item "-m"
Forces the use of a random multi-cast \s-1MAC\s0 address when generating
version 1 UUIDs. By default the real physical \s-1MAC\s0 address of the system
is used.
.IP "\fB\-n\fR \fIcount\fR" 3
.IX Item "-n count"
Generate \fIcount\fR UUIDs instead of just a single one (the default).
.IP "\fB\-1\fR" 3
.IX Item "-1"
If option \fB\-n\fR is used with a \fIcount\fR greater than \f(CW1\fR, then this
option can enforce the reset the \s-1UUID\s0 context for each generated \s-1UUID\s0.
This makes no difference for \fIversion\fR \f(CW3\fR, \f(CW4\fR and \f(CW5\fR UUIDs. But
version \f(CW1\fR UUIDs are based on the previously generated \s-1UUID\s0 which is
remembered in the \s-1UUID\s0 context of the \s-1API\s0. Option \fB\-1\fR deletes the
remembered \s-1UUID\s0 on each iteration.
.IP "\fB\-F\fR \fIformat\fR" 3
.IX Item "-F format"
Representation format for importing or exporting an \s-1UUID\s0. The
following (case insensitive) format identifiers are currently recognized:
.RS 3
.ie n .IP "\*(C`BIN\*(C' (binary representation)" 4
.el .IP "\f(CW\*(C`BIN\*(C'\fR (binary representation)" 4
.IX Item "BIN (binary representation)"
This is the raw 128 bit network byte order binary representation of a
\&\s-1UUID\s0. Example is the octet stream \f(CW\*(C`0xF8 0x1D 0x4F 0xAE 0x7D 0xEC 0x11
0xD0 0xA7 0x65 0x00 0xA0 0xC9 0x1E 0x6B 0xF6\*(C'\fR.
.ie n .IP "\*(C`STR\*(C' (string representation)" 4
.el .IP "\f(CW\*(C`STR\*(C'\fR (string representation)" 4
.IX Item "STR (string representation)"
This is the 36 character hexadecimal \s-1ASCII\s0 string representation of a
\&\s-1UUID\s0. Example is the string "\f(CW\*(C`f81d4fae\-7dec\-11d0\-a765\-00a0c91e6bf6\*(C'\fR".
.ie n .IP "\*(C`SIV\*(C' (single integer value representation)" 4
.el .IP "\f(CW\*(C`SIV\*(C'\fR (single integer value representation)" 4
.IX Item "SIV (single integer value representation)"
This is the maximum 39 character long single integer
value representation of a \s-1UUID\s0. Example is the string
"\f(CW329800735698586629295641978511506172918\fR".
.RE
.RS 3
.RE
.IP "\fB\-o\fR \fIfilename\fR" 3
.IX Item "-o filename"
Write output to \fIfilename\fR instead of to \fIstdout\fR.
.IP "\fB\-d\fR" 3
.IX Item "-d"
Decode a given \s-1UUID\s0 (given as a command line argument or if the command
line argument is "\f(CW\*(C`\-\*(C'\fR" the \s-1UUID\s0 is read from \fIstdin\fR) and dump textual
information about the \s-1UUID\s0.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.Vb 3
\& # generate DCE 1.1 v1 UUID (time and node based)
\& $ uuid \-v1
\& 01c47915\-4777\-11d8\-bc70\-0090272ff725
\&
\& # decode and dump DCE 1.1 v1 UUID (time and node based)
\& $ uuid \-d 01c47915\-4777\-11d8\-bc70\-0090272ff725
\& encode: STR:     01c47915\-4777\-11d8\-bc70\-0090272ff725
\&         SIV:     2349374037528578887923094374772111141
\& decode: variant: DCE 1.1, ISO/IEC 11578:1996
\&         version: 1 (time and node based)
\&         content: time:  2004\-01\-15 16:22:26.376322.1 UTC
\&                  clock: 15472 (usually random)
\&                  node:  00:90:27:2f:f7:25 (global unicast)
\&
\& # generate DCE 1.1 v3 UUID (name based)
\& $ uuid \-v3 ns:URL http://www.ossp.org/
\& 02d9e6d5\-9467\-382e\-8f9b\-9300a64ac3cd
\&
\& # decode and dump DCE 1.1 v3 UUID (name based)
\& $ uuid \-d 02d9e6d5\-9467\-382e\-8f9b\-9300a64ac3cd
\& encode: STR:     02d9e6d5\-9467\-382e\-8f9b\-9300a64ac3cd
\&         SIV:     3789866285607910888100818383505376205
\& decode: variant: DCE 1.1, ISO/IEC 11578:1996
\&         version: 3 (name based, MD5)
\&         content: 02:D9:E6:D5:94:67:08:2E:0F:9B:93:00:A6:4A:C3:CD
\&                  (not decipherable: MD5 message digest only)
\&
\& # generate DCE 1.1 v4 UUID 4 (random data based)
\& $ uuid \-v4
\& eb424026\-6f54\-4ef8\-a4d0\-bb658a1fc6cf
\&
\& # decode and dump DCE 1.1 v4 UUID 4 (random data based)
\& $ uuid \-d eb424026\-6f54\-4ef8\-a4d0\-bb658a1fc6cf
\& encode: STR:     eb424026\-6f54\-4ef8\-a4d0\-bb658a1fc6cf
\&         SIV:     312712571721458096795100956955942831823
\& decode: variant: DCE 1.1, ISO/IEC 11578:1996
\&         version: 4 (random data based)
\&         content: EB:42:40:26:6F:54:0E:F8:24:D0:BB:65:8A:1F:C6:CF
\&                  (no semantics: random data only)
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIuuid\fR\|(3), \fIOSSP::uuid\fR\|(3).


libs/uuid/php/uuid.ts
<?php
##
##  OSSP uuid - Universally Unique Identifier
##  Copyright (c) 2004-2007 Ralf S. Engelschall <rse@engelschall.com>
##  Copyright (c) 2004-2007 The OSSP Project <http://www.ossp.org/>
##
##  This file is part of OSSP uuid, a library for the generation
##  of UUIDs which can found at http://www.ossp.org/pkg/lib/uuid/
##
##  Permission to use, copy, modify, and distribute this software for
##  any purpose with or without fee is hereby granted, provided that
##  the above copyright notice and this permission notice appear in all
##  copies.
##
##  THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED
##  WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
##  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
##  IN NO EVENT SHALL THE AUTHORS AND COPYRIGHT HOLDERS AND THEIR
##  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
##  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
##  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
##  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
##  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
##  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
##  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
##  SUCH DAMAGE.
##
##  uuid.ts: PHP/Zend API test procedure (language: php)
##

##
##  INITIALIZATION
##

$php_version = $argv[1];

print "++ loading DSO uuid.so (low-level API)\n";
if (!extension_loaded('uuid')) {
    dl('modules/uuid.so');
}

print "++ loading PHP uuid.php${php_version} (high-level API)\n";
require "uuid.php${php_version}";

print "++ establishing assertion environment\n";
assert_options(ASSERT_ACTIVE, 1);
assert_options(ASSERT_WARNING, 0);
assert_options(ASSERT_QUIET_EVAL, 1);
function my_assert_handler($file, $line, $code)
{
    echo "ASSERTION FAILED: $file: $line: $code\n";
    exit(1);
}
assert_options(ASSERT_CALLBACK, 'my_assert_handler');

##
##  LOW-LEVEL API TESTING
##

print "++ testing low-level C-style API:\n";

$uuid = 42;
$rc = uuid_create(&$uuid);
assert('$rc == 0');
assert('$uuid != 42');

$rc = uuid_make($uuid, UUID_MAKE_V1);
assert('$rc == 0');

$str = "foo";
$rc = uuid_export($uuid, UUID_FMT_STR, &$str);
assert('$rc == 0');
assert('$str != "foo"');
print "UUID: $str\n";

$uuid_ns = 42;
$rc = uuid_create(&$uuid_ns);
assert('$rc == 0');

$rc = uuid_load($uuid_ns, "ns:URL");
assert('$rc == 0');

$rc = uuid_make($uuid, UUID_MAKE_V3, $uuid_ns, "http://www.ossp.org/");
assert('$rc == 0');

$str = "bar";
$rc = uuid_export($uuid, UUID_FMT_STR, &$str);
assert('$rc == 0');
assert('$str != "bar"');
#assert('$str == "02d9e6d5-9467-382e-8f9b-9300a64ac3cd"');
print "UUID: $str\n";

$rc = uuid_destroy($uuid);
assert('$rc == 0');

$rc = uuid_create(&$uuid);
assert('$rc == 0');

$rc = uuid_import($uuid, UUID_FMT_STR, $str);
assert('$rc == 0');

$str = "baz";
$rc = uuid_export($uuid, UUID_FMT_STR, &$str);
assert('$rc == 0');
assert('$str != "baz"');
#assert('$str == "02d9e6d5-9467-382e-8f9b-9300a64ac3cd"');
print "UUID: $str\n";

$clone = null;
$rc = uuid_clone($uuid, &$clone);
assert('$rc == 0');
assert('$clone != null');

$rc = uuid_destroy($uuid);
assert('$rc == 0');

$str = "quux";
$rc = uuid_export($clone, UUID_FMT_STR, &$str);
assert('$rc == 0');
assert('$str != "quux"');
#assert('$str == "02d9e6d5-9467-382e-8f9b-9300a64ac3cd"');
print "UUID: $str\n";

##
##  HIGH-LEVEL API TESTING
##

print "++ testing high-level OO-style API:\n";

$uuid = new UUID;
$uuid->make(UUID_MAKE_V1);
$str = $uuid->export(UUID_FMT_STR);
print "UUID: $str\n";

$uuid_ns = new UUID;
$uuid_ns->load("ns:URL");
$uuid->make(UUID_MAKE_V3, $uuid_ns, "http://www.ossp.org/");
$str = $uuid->export(UUID_FMT_STR);
print "UUID: $str\n";
$uuid = null;
$uuid_ns = null;

$uuid = new UUID;
$uuid->import(UUID_FMT_STR, $str);
$str = $uuid->export(UUID_FMT_STR);
print "UUID: $str\n";

if ($php_version == 4) {
    eval('$clone = $uuid->clone();');
}
else {
    eval('$clone = clone $uuid;');
}
$uuid = null;

$str = $clone->export(UUID_FMT_STR);
print "UUID: $str\n";

$clone = null;

?>
